api/utilities.py

```diff
@@ -137,3 +147,159 @@ async def add_parameter_name_and_datetime(request, parameter_name: str | None, d
+    """
+    Function for getting the z values from the z parameter.
+    """
+    # it can be z=value1,value2,value3: z=2,10,80
+    # or z=minimum value/maximum value: z=10/100
+    # or z=Rn/min height/height interval: z=R20/100/50
+    try:
+        split_on_slash = z.split("/")
+        if len(split_on_slash) == 2:
+            z_min = float(split_on_slash[0]) if split_on_slash[0] != ".." else float("-inf")
+            z_max = float(split_on_slash[1]) if split_on_slash[1] != ".." else float("inf")
+            return [z_min, z_max]
+        elif len(split_on_slash) > 2:
+            return get_z_values_from_interval(split_on_slash)
+        else:
+            return list(map(float, z.split(",")))
+    except ValueError:
+        raise HTTPException(status_code=400, detail=f"Invalid levels value: {z}")
+
+
+def get_z_values_from_interval(interval: list[str]) -> list[float]:
+    """
+    Function for getting the z values from a repeating-interval pattern.
+    """
+    # Make sure the pattern is Rn/n/n
+    # Allow optional decimals in interval and starting value
+    # TODO: Can the starting value be negative? Can the interval be negative to go in reverse?
+    pattern = re.compile(r"^R\d+/\d+(\.\d+)?/\d+(\.\d+)?$")
+    if not pattern.match("/".join(interval)):
+        raise HTTPException(status_code=400, detail=f"Invalid levels repeating-interval: {'/'.join(interval)}")
+
+    amount_of_intervals = int(interval[0][1:])
+    interval_value = float(interval[1])
+    starting_value = float(interval[2])
+
+    # Round to 3 decimals to avoid floating point errors
+    return [round(starting_value + i * interval_value, 3) for i in range(amount_of_intervals)]
+
+
+def is_float(element: any) -> bool:
+    if element is None:
+        return False
+    try:
+        float(element)
+        return True
+    except ValueError:
+        return False
+
+
+async def get_unique_values_for_metadata(field: str) -> list[str]:
+    request = dstore.GetTSAGRequest(attrs=[field])
+    response = await get_ts_ag_request(request)
+    return [getattr(i.combo, field) for i in response.groups]
+
+
+def filter_observations_for_z(observations, z):
+    if z:
+        z_values = get_z_levels_or_range(z)
+        if len(z.split("/")) == 2:
+            z_min, z_max = z_values[0], z_values[1]
+            observations = [
+                obs
+                for obs in observations
+                if is_float(obs.ts_mdata.level) and z_min <= float(obs.ts_mdata.level) <= z_max
+            ]
+        else:
+            observations = [
+                obs for obs in observations if is_float(obs.ts_mdata.level) and float(obs.ts_mdata.level) in z_values
+            ]
+    return observations
+
+
+def numeric_sort_key(value: str) -> float:
+    """
+    Converts a string to a float for comparison, returns infinity if the string is not convertible.
+    """
+    try:
+        return float(value)
+    except ValueError:
+        return float("inf")
```

Comment:
Would NaN be better, or do you specifically need something big?

Reply:
I was under the impression that using NaN in sorting could have some unexpected behavior, so I decided to use positive `inf` to push non float values to the end of the list. The ingestor *should* take care that there are no `levels` in the system which are not castable to `float`, so this was just a *"better safe than sorry"* solution.